### PR TITLE
FIP-0045: Implement migrations

### LINKED ISSUE #69: FIP-0045: Migration
Here is the proposed  workflow for the FIP-0045 migration:

- [ ] Step 0: Migrate the MinerActors (happens as part of the broader migration)
  - SectorOnChainInfos are updated in the Sectors AMT as well as SectorsSnapshot AMTs in each Deadline
- [ ] Step 1: Create the DataCap actor, and put it into the state tree
  - This only needs the v8 VerifiedRegistry's VerifiedClients field as input, with tokens being granted to each client
-  [ ] Step 2: Migrate the VerifregActor
   - The input needed is a `map[DealID]DealProposal` of verified deals in the PendingProposals Set
   - This can be constructed from the v8 market state, and provided as input
   - For each of these deals, create an Allocation in the Allocations HAMT
   - Return a `map[DealID]AllocationID` for input in the next step

- [ ] Step 3: Migrate the MarketActor 
  - This requires the `map` result of the previous step (the Verifreg migration) as input
  - This map effectively becomes the `PendingDealAllocationIds`

### PR DESCRIPTION
**Pull request by arajasek**
Closes #69

**The sequence of steps is:**

- Step 0: Add the DataCap actor to the v8 state tree with a fake Head
- Step 1A: Migrate all actors except the StorageMarketActor and VerifiedRegistryActor as part of the "regular" migration workflow (more details on the migrations below), set the results in the output state tree
- Step 1B: In a separate, parallel goroutine migrate the VerifiedRegistryActor, and then the StorageMarketActor
- Block on 1A and 1B
- Step 2: Set the VerifiedRegistryActor and StorageMarketActor in the output state tree
- Step 3: Flush and return

**Miner Migration**

The minerMigrator performs the following migrations:
- FIP-0029: Sets the `Beneficary` to the `Owner`, and sets empty values for `BeneficiaryTerm and PendingBeneficiaryTerm`
- FIP-0034: For each `SectorPreCommitOnChainInfo` in `PreCommitedSectors`, calculates the unsealed CID (assuming there are deals)
- FIP-0045: For each SectorOnChainInfo in Sectors, set `SimpleQAPower = (DealWeight == 0 && VerifiedDealWeight == 0)`
- FIP-0045: For each Deadline in Deadlines: for each `SectorOnChainInfo` in `SectorsSnapshot`, set `SimpleQAPower = (DealWeight == 0 && VerifiedDealWeight == 0)`

The migration has the following layers of caching:

- at the `ActorHead` level -- if a matching `(address, Head)` tuple is found in the cache, use the cached value
- at the `SectorsAmt` level -- if a matching `Sectors` CID is found in the cache, use the cached value
  - This is used for migrating both the `Sectors` AMT, as well as the `SectorsSnapshot` AMTs in each deadline
- at the "previous SectorsAmt" level -- if we have the previous input and output of migrating this particular miner's SectorsAmt before, load them, and transform previous output to current output based on the diff between previous input and current input

Further caching is possible at the following levels;
- at the `PreCommits` level -- if a matching `PreCommittedSectors` CID is found in the cache, use the cached value
- at the individual `PreCommit` level -- if a matching `SectorPreCommitOnChainInfo` CID is found in the cache, use the cached value

**DataCap "migration"**

The `DataCap` State is set based on the Verified Registry's `VerifiedClients` map. For each client:
- its balance is converted to tokens that are added to the `TokenState`
- an entry is created in the TokenState's `allowances` map, with infinite allowance and the Market actor as operator

**VerifiedRegistry & Market migration**

As in #69

### CODE REVIEW

File: builtin/v8/util/adt/set.go
Comment (ZenGround0) [inline builtin/v8/util/adt/set.go:8]: nit: Since this is code and not just structs it would be nice to see a link to where this is copied from (if it is copied)

File: builtin/v9/migration/top.go
Comment (ZenGround0) [inline builtin/v9/migration/top.go:355]: For clarity if possible please split into three different migration functions with inputs and outputs.  From talking offline it sounds like this should be possible.

File: builtin/v9/migration/miner.go
Comment (ZenGround0) [inline builtin/v9/migration/miner.go:110]: Potentially worth caching

File: builtin/v9/migration/miner.go
Comment (ZenGround0) [inline builtin/v9/migration/miner.go:168]: nit: lets use consistent naming between newSectors and deadlinesOut

File: builtin/v9/migration/miner.go
Comment (ZenGround0) [inline builtin/v9/migration/miner.go:201]: Nice let's make sure we test this

File: builtin/v9/migration/miner.go
Comment (ZenGround0) [inline builtin/v9/migration/miner.go:356]: Is this now addressed with the cache.Load()?

File: builtin/v9/migration/miner.go
Comment (ZenGround0) [inline builtin/v9/migration/miner.go:333]: We probably don't want to swallow these errors.

File: builtin/v9/migration/miner.go
Comment (ZenGround0) [inline builtin/v9/migration/miner.go:278]: I'm not sure if I'm reading it right but [Diff](https://github.com/filecoin-project/go-amt-ipld/blob/723fff1300c57c2f866505a7f3bd0796817e8bbf/diff.go#L42) doesn't seem to quickly check that the cids are equal and return immediately.  If this is right it is important for us to check equality quickly and set outArray to prevOutArray.  It might speed us up a lot on second run.

File: builtin/v9/migration/miner.go
Comment (ZenGround0) [inline builtin/v9/migration/miner.go:325]: super nit: delete new line

File: builtin/v9/migration/miner.go
Comment (ZenGround0) [inline builtin/v9/migration/miner.go:378]: If we are doing diffing in the SectorInfo context it probably makes sense to do diffing here too.  The function inside load should then be the extracted function used above.

File: builtin/v9/migration/miner.go
Comment (ZenGround0) [reply builtin/v9/migration/miner.go:278]: Aha I'm missingt that this is inside cache.Load

File: manifest/manifest.go
Comment (geoff-vball) [inline manifest/manifest.go:46]: We have a version on this in `chain/actors/manifest.go`. I think this is a better location for this, but it still should take actors version as an input.

File: builtin/v8/util/adt/set.go
Comment (arajasek) [reply builtin/v8/util/adt/set.go:8]: Yessir!

File: builtin/v9/migration/top.go
Comment (arajasek) [reply builtin/v9/migration/top.go:355]: Yessir!

File: builtin/v9/migration/miner.go
Comment (ZenGround0) [reply builtin/v9/migration/miner.go:378]: And to be more clear I'm suggesting diffing against `MinerPrevSectorsInKey(minerAddr)`. We're never going to be more than 1 Day plus Wait time between premigrations of diffs.

If things are still taking a long time and you are feeling experimental you could introduce deadline index / address keys to do diffing against previously cached deadline sectors snapshots but I guess that this won't be a good complexity performance tradeoff.

File: builtin/v9/migration/miner.go
Comment (arajasek) [reply builtin/v9/migration/miner.go:110]: Agreed. I'm gonna start an issue capturing possible further caching / speedups. I do want to keep the cache focused on highest impact stuff, because it can get big...

File: builtin/v9/migration/miner.go
Comment (arajasek) [reply builtin/v9/migration/miner.go:168]: yeeee

File: builtin/v9/migration/miner.go
Comment (arajasek) [reply builtin/v9/migration/miner.go:333]: I was torn about this, but I think I agree. Or at least don't feel strongly about it.

File: builtin/v9/migration/miner.go
Comment (arajasek) [reply builtin/v9/migration/miner.go:378]: It's a little bit more complicated than that. The diffing for the Sectors AMT is based on snapshotting (for each miner) the input and output tree at the time of the premigration.

It's unclear what the equivalent to do here would be. We _could_ snapshot each deadline for each miner, but that feels like a lot. We could also just use the input/output of the migration of the "real" Sectors AMT (which we perform right before we start migrating deadlines) as the basis of the diff for the sectors snapshots in each of the deadlines. Realistically, that's as good a baseline for the snapshotted AMTs.

My thought was that all but 3-4 of the deadlines per miner will be perfectly cached by the premigration anyway (since their SectorsSnapshots can't change in the 2 hours between the premigration and the migration). Further, if they do change, there's a good chance the snapshot winds up being the same as the "real" Sectors AMT, and so are found in the cache from the previous step.

All that to say, I agree that some caching is possible here, but there's some open questions about the best way to do it. Some part of me is wary of introducing more caching than necessary lest doing so introduce bugs.

Comment (ZenGround0) [reply builtin/v9/migration/miner.go:381]: > We could also just use the input/output of the migration of the "real" Sectors AMT (which we perform right before we start migrating deadlines) as the basis of the diff for the sectors snapshots in each of the deadlines. Realistically, that's as good a baseline for the snapshotted AMTs.

Yeah I think this is what I was trying to suggest with " diffing against MinerPrevSectorsInKey(minerAddr)" thought I'm missing some details.  Good point that "all but 3-4 of the deadlines per miner will be perfectly cached by the premigration anyway" so doing more caching is definitely not worth it.

File: builtin/v9/migration/verifreg.go
Comment (ZenGround0) [inline builtin/v9/migration/verifreg.go:88]: I don't think this adt.Map ever gets added to the allocationsMapMap golang map (https://go.dev/play/p/ouAvN7Ea813) causing all but the last iterated verified deal from client to be added.

**Review by ZenGround0 — COMMENTED** [review]
A couple things before looking at top level

File: builtin/v9/migration/miner.go
Comment (arajasek) [reply builtin/v9/migration/miner.go:265]: That...makes complete sense, yes.

File: builtin/v9/migration/miner.go
Comment (arajasek) [reply builtin/v9/migration/miner.go:381]: Yeah, I'd prefer to leave that to a followup if that works for you -- see https://github.com/filecoin-project/go-state-types/issues/69.

File: builtin/v9/migration/top.go
Comment (ZenGround0) [inline builtin/v9/migration/top.go:183]: nit: move comment to code it refers to

File: builtin/v9/migration/top.go
Comment (ZenGround0) [reply builtin/v9/migration/top.go:347]: For threadsafety without hacking into the main wait group logic you can read inputs from actorsIn before spawning goroutine and write to actorsOut after all other migrations have finished.

File: builtin/v9/verifreg/verifreg_types.go
Comment (ZenGround0) [inline builtin/v9/verifreg/verifreg_types.go:87]: Yeah what happens when we delete this?

**Review by ZenGround0 — APPROVED** [review]
I'm ok with this merging once correctness is addressed and defer performance to later.  However I would prefer getting at least the deferred migrations in parallel with the rest before merge since this looks like a great opportunity and in need of testing asap.

File: builtin/v9/migration/miner.go
Comment (arajasek) [reply builtin/v9/migration/miner.go:265]: Ah, the limitation is that the cache (as currently written) can only have `value` of type `cid` (not `{cid, cid}`.

File: builtin/v9/migration/verifreg.go
Comment (arajasek) [reply builtin/v9/migration/verifreg.go:100]: Hmm, I implemented this based on the FIP https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0045.md#migration. Not sure what to do.

File: builtin/v9/migration/verifreg.go
Comment (arajasek) [reply builtin/v9/migration/verifreg.go:88]: GOOD catch, thank you.

File: builtin/v9/migration/top.go
Comment (arajasek) [reply builtin/v9/migration/top.go:183]: This was intentionally placed, because the market state loading is for The Miner Actor. I can move it if it's only serving to confuse, though.

File: builtin/v9/migration/verifreg.go
Comment (ZenGround0) [reply builtin/v9/migration/verifreg.go:100]: Offline I said its ok to keep as is.  One small change: I think you should set TermMax to DealMaxDuration + MarketDefaultAllocationTermBuffer.  This way deals with maximal deal lifetime get the same TermMax as they would from market code.

File: builtin/v9/migration/top.go
Comment (ZenGround0) [reply builtin/v9/migration/top.go:183]: ah ok in that case writing a bit more would also do it

File: builtin/v9/migration/miner.go
Comment (ZenGround0) [reply builtin/v9/migration/miner.go:265]: Ah in that case this now makes complete sense

File: builtin/v9/migration/top.go
Comment (arajasek) [reply builtin/v9/migration/top.go]: Done

File: builtin/v9/migration/verifreg.go
Comment (anorth) [inline builtin/v9/migration/verifreg.go:34]: This is a large map (every deal currently active). It would be much faster to iterate PendingProposals and look them up in Proposals.

File: builtin/v9/migration/verifreg.go
Comment (anorth) [inline builtin/v9/migration/verifreg.go:133]: ‼️ this should be `nextAllocationId`.

File: builtin/v9/migration/verifreg.go
Comment (anorth) [inline builtin/v9/migration/verifreg.go:70]: Nit: there'd be good clarity gains from extracting pedestrian logic like this into functions so that the main migration code gets straight to the point.

File: builtin/v9/migration/verifreg.go
Comment (anorth) [inline builtin/v9/migration/verifreg.go:102]: ‼️ The max term should be a policy value _more than_ the deal duration, not a fixed value for all deals.

File: builtin/v9/verifreg/verified_registry_state.go
Comment (anorth) [inline builtin/v9/verifreg/verified_registry_state.go:108]: If the state somehow ended up like this, I don't think this is the right place to catch it. This does not mirror the Rust code.

File: builtin/v9/verifreg/verified_registry_state.go
Comment (anorth) [inline builtin/v9/verifreg/verified_registry_state.go:142]: Ditto. Less smarts in the state layer.

File: builtin/v9/verifreg/verified_registry_state.go
Comment (anorth) [inline builtin/v9/verifreg/verified_registry_state.go:163]: Nit: suggest AllocationsAsMap, or LoadAllocationsToMap or something else that suggests the amount of work this will do

**Review by anorth — CHANGES_REQUESTED** [review]
I reviewed primarily for correctness of the state migrations themselves. I'm pretty far from the mechanics of how they are executed now and did not dive deep there.

File: builtin/v9/migration/verifreg.go
Comment (arajasek) [reply builtin/v9/migration/verifreg.go:102]: @anorth I based this on the [text of the FIP ](https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0045.md#migration). Am I misunderstanding something here?

Comment (anorth) [reply builtin/v9/migration/verifreg.go:108]: You're right, I take it back.

Comment (arajasek) [reply builtin/v9/migration/verifreg.go:108]: Zen had a slightly different suggestion [here](https://github.com/filecoin-project/go-state-types/pull/85#discussion_r988251213) -- do you have an opinion?

File: builtin/shared.go
Comment (arajasek) [reply builtin/shared.go:13]: @anorth It's also needed at other times when the datacap actor state is constructed, eg. when creating new genesis blocks, starting directly from v9 actors.

File: builtin/v9/verifreg/verified_registry_state.go
Comment (arajasek) [reply builtin/v9/verifreg/verified_registry_state.go:27]: Good call.

File: builtin/v9/migration/verifreg.go
Comment (anorth) [reply builtin/v9/migration/verifreg.go:108]: Wait, yeah I un-take it back. We want the same behaviour as the market would have done. If a client set 540d deal term, the max should be +buffer on that.

Comment (arajasek) [reply builtin/v9/migration/verifreg.go:108]: Good, thanks! I will implement this change and update the FIP for overall Zen.

File: builtin/v9/migration/verifreg.go
Comment (anorth) [reply builtin/v9/migration/verifreg.go:35]: Oh right, I see. There's good reason for the set of CIDs.

File: builtin/v9/migration/verifreg.go
Comment (arajasek) [reply builtin/v9/migration/verifreg.go:139]: Holy fork, thank you!

File: builtin/v9/verifreg/verified_registry_state.go
Comment (arajasek) [reply builtin/v9/verifreg/verified_registry_state.go:108]: I'd prefer to leave it -- I think it's reasonable for the state layer to know that if it's asked to find a claim for a specific provider, then the returned claim should match that provider.

No harm done in the sanity check, I think.

File: builtin/v9/migration/verifreg.go
Comment (anorth) [inline builtin/v9/migration/verifreg.go:90]: I think we want `proposal.Duration() + market9.MarketDefaultAllocationTermBuffer`

**Review by anorth — APPROVED** [review]

**Review by ZenGround0 — APPROVED** [review]

**Pull request merged**
